Code/GraphMol/RGroupDecomposition/RGroupDecompData.h

```diff
@@ -457,26 +460,35 @@ struct RGroupDecompData {
         best_score = newscore;
         best_permutation = iterator.permutation;
       }
-      count++;
+      checkForTimeout(t0, params.timeout);
     }
 
     if (ties.size() > 1) {
-      // choose one that doesn't add rgroups inappropriately
-      //  an rgroup is added when
-      //  (1) the label is <=0
-      //  (2) the group has any substituent with heavy atoms
-      //   XXX Might be more efficient to add this comp to the score function
-      int smallest_added_rgroups = 100000000;
-      for (auto tied_permutation : ties) {
-        int num_added_rgroups = compute_num_added_rgroups(tied_permutation);
+      size_t min_perm_value = 0;
+      size_t smallest_added_rgroups = labels.size();
+      for (const auto &tied_permutation : ties) {
+        std::vector<int> heavy_counts;
+        std::vector<int> largest_heavy_counts(labels.size(), 0);
+        size_t num_added_rgroups =
+            compute_heavy_rgroup_counts(tied_permutation, heavy_counts);
         if (num_added_rgroups < smallest_added_rgroups) {
           smallest_added_rgroups = num_added_rgroups;
           best_permutation = tied_permutation;
+        } else if (num_added_rgroups == smallest_added_rgroups) {
+          if (heavy_counts < largest_heavy_counts) {
```

Comment:
This vector comparison is unclear to me, I think it needs an explanation.  It appears to be biasing toward "early" rgroups which in this case might be the non user defined ones.  I think this is potentially a mistake.

Reply:
See above explanation.